website/hw/hw4.rst

```diff
@@ -180,103 +218,389 @@ As you are stepping through your program, you might want to be able to
+you may want to set breakpoints close to the point where the segfault happens,
+to see exactly what happens in the lines leading up to the segfault. However,
+to do this, you must first use the ``kill`` to stop the current run of the program
+(without exiting GDB).
+You will then be able to ``run`` the program again, stopping at any additional
+breakpoints you specify.
 
--  Where was the segfault? Give the function, line number, and
-   copy-paste the line of code itself. What was causing the segfault and
-   how did you fix it?
+Try running the ``segfault`` program with gdb and using the above
+commands to figure out what is causing the segfault. You will find
+the answer at the bottom of this homework.
 
-It is a good idea to fix one error before moving on to the next. Make
-sure to also commit and push your fix for each segfault!
+Task 2: Fixing Segfaults in the Micro Editor
+--------------------------------------------
 
-You must also answer the following question for the entire task:
+(XX points)
 
--  Describe the steps you took in gdb to narrow down the exact location
-   of one of the segfaults. You can also copy-paste the contents of one
-   debugging session with gdb, with a brief note at the end explaining
-   how that debugging session narrowed down the location of the error.
+TODO: Micro will segfault under certain conditions. Figure out what the issue is.
 
-Task 4: Logic Error Debugging
+
+Task 3: Logic Error Debugging
 -----------------------------
 
-(20 points)
+(XX points)
 
 Logic errors can happen even when your program compiles and runs, but
 may result in incorrect behavior. The commands we went through in the
 warmup can help you find logic errors by allowing actions such as
 stepping through the program line-by-line, printing variables, and
 seeing when variables are changed.
 
-Compile ``task4.c`` and run it. As you can see from the print messages,
-the variables don’t have the expected values! Use gdb to find the five
-errors. For each error you find, answer the following questions on
-Gradescope:
-
--  Copy-paste the relevant lines of code, explain what is wrong with
-   them, and then copy-paste the fixed code (with a brief explanation of
-   how you fixed the code).
-
-Please note that, while you might also be able to find the logic errors
-just by looking at the code, that will rob you of the opportunity to
-practice using gdb to step through a running program.
+TODO: Finding a logic error in Micro editor
 
 Submitting your lab
 -------------------
 
 Please note that you will not be submitting your code through
-Gradescope. Instead, make sure that the questions posed above are
+Gradescope. Instead, make sure that the questions posed in each task are
 answered on Gradescope. That said, we still need you to push your code
 in case we need to look at any of your code (but we will not be grading
 the code itself).
+
+
+Answers to Practice Questions
+-----------------------------
+
```

Comment:
It may also be helpful to number the practice questions, and include in this section a quick answer key in addition to the walkthrough solution (so that students who think they understood can quickly check their answers)